system/masking/include/datamaskingplugin.hpp

```diff
@@ -0,0 +1,2072 @@
+        }
+    protected:
+        typedef char* (*Finder)(const char*, const char*);
+        //using Finder = std::function<char*(const char*, const char*)>;
+        std::string startToken;
+        std::string endToken;
+        Finder      finder = nullptr;
+    public:
+        using CRule::CRule;
+        virtual bool applyRule(char* buffer, size_t length, const IDataMaskingProfileContext* context) const
+        {
+            if (isEmptyString(buffer) || 0 == length)
+                return false;
+            if (std::numeric_limits<size_t>::max() == length || '\0' == buffer[length])
+            {
+                return applyRule(buffer, context);
+            }
+            else
+            {
+                Terminator terminator(&buffer[length]);
+                return applyRule(buffer, context);
+            }
+        }
+    protected:
+        virtual bool configureTokens(const IPTree& configuration)
+        {
+            bool invalidated = false;
+            const char* tmp = configuration.queryProp("@start-token");
+            if (isEmptyString(tmp))
+            {
+                trace().uerrlog("%s: missing required start token", tracePrefix());
+                invalidated = true;
+            }
+            else
+                startToken = tmp;
+            tmp = configuration.queryProp("@end-token");
+            if (isEmptyString(tmp))
+            {
+                trace().log(MCuserInfo, "%s: missing end token; assuming EOLN", tracePrefix());
+                endToken = "\n";
+            }
+            else
+                endToken = tmp;
+            return !invalidated;
+        }
+        virtual bool configureCase(const IPTree& configuration)
+        {
+            if (configuration.getPropBool("@match-case", false))
+                finder = strstr_sensitive;
+            else
+                finder = strstr_insensitive;
+            return true;
+        }
+        virtual bool applyRule(char* buffer, const IDataMaskingProfileContext* context) const
+        {
+            bool applied = false;
+            char* start = finder(buffer, startToken.c_str());
+            char* end;
+            while (start)
+            {
+                start += startToken.length();
+                end = finder(start, endToken.c_str());
+                if (end)
+                {
+                    if (valueType.applyMask(nullptr, start, 0, end - start, context))
+                        applied = true;
+                    start = finder(end + endToken.length(), startToken.c_str());
+                }
+                else
+                    start = nullptr;
+            }
+            return applied;
+        }
+    private:
+        inline static char* strstr_sensitive(const char* haystack, const char* needle)
+        {
+            return const_cast<char*>(strstr(haystack, needle));
+        }
+        inline static char* strstr_insensitive(const char* haystack, const char* needle)
```

Comment:
There is an existing implementation named "stristr" which I think also only currently works for ascii.  Could common the functions up with whatever implementation is most efficient.  When going beyond ascii I agree, I would probably focus on utf8.  There is a local aware toupper although the wrapper seems to do a dynamic cast and I would probably unwind the wrapper a bit and make sure I avoided that.  Would still want to look at the efficiency of toupper compared to a homegrown table like the one in stristr.

Reply:
I saw that, but wanted to avoid requiring a system utility to link to an ESP library. I tried strcasestr, which worked in Linux but broke the Windows builds.